### PR TITLE
[release/6.0] Reduce net core app current package dependencies, increase direct update availability

### PR DESCRIPTION
**Pull request by ericstj**
## Customer Impact

- [x] Customer reported
- [x] Found internally

Customers using packages like `Microsoft.Extensions.Logging.Console`, `System.Memory.Data`, `Microsoft.Extensions.Hosting`, etc and targeting `net6.0` need to update package references for packages like `System.Text.Json`, `System.Text.Encodings.Web`, `System.Collections.Immutable`, etc even though those libraries are provided by the shared framework.

We can reduce the number of places where folks need to update packages by omitting packages when the same library is provided by the framework. 

Additionally we've shipped many packages over the course of `net6.0` but haven't shipped every package that depends on them - causing customers to update more transitive dependencies to get updates.  We have seen slower up-take for security fixes when transitive updates are required - the goal here is to ensure more direct updates are available.

## Regression

- [ ] Yes
- [X] No

## Testing

Build packages.  Add validation to make sure all up-stack packages ship as well.   Examined package diffs.

## Risk

Low.  This is removing package references for a few packages and enabling more packages.  The biggest risk here is that we'll be enabling a lot of packages that need to flow in servicing - a total of 55 - but the goal here is that this set of packages will help customers better get clean of CVEs.

## Background

This drops package dependencies from all packages which can reference the framework copy of the same library (without downgrading the library exposed to a compatible TFM like netstandard2.0).

I also added a feature to our build that enforces transitive servicing.  You can specify `ServiceTransitiveDependencies` to make sure that you enable all up-stack packages for shipping when enabling a single package.

I enabled all packages I changed, then all upstack packages (separate commits).

This should improve the situation where folks are asked to update just to update a package reference on the latest framework.

It will also help reduce application size since the libraries will no longer be bundled in the app.

## PR detail

This is a backport of https://github.com/dotnet/runtime/pull/107161.

I also reviewed all packages that ever shipped in 6.0 with a CVE to ensure that we've shipped at least one package version that references that - to ensure folks have availability of a direct update for the CVE.

The only "code changes" in this PR are to remove dependencies when the same dependency is in the shared framework.  You can see the result of that diff here: https://github.com/ericstj/diffs/commit/499919a93feb8f6c6ac7474e975e3f4f814871a4

The remaining changes are infrastructure or enabling packages.

### CODE REVIEW

File: src/libraries/System.Memory.Data/src/System.Memory.Data.csproj [hunk 1]
```diff
@@ -4,8 +4,8 @@
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <Nullable>enable</Nullable>
     <IsPackable>true</IsPackable>
```
Comment (carlossanlop) [inline src/libraries/System.Memory.Data/src/System.Memory.Data.csproj:6]: Huh, why did this package never showed errors when publishing the nupkgs? It's been turned on all this time but we never saw versioning conflicts?

File: src/libraries/Microsoft.Extensions.Http/src/Microsoft.Extensions.Http.csproj [hunk 2]
```diff
@@ -8,6 +8,8 @@
 
 Commonly Used Types:
 System.Net.Http.IHttpClientFactory</PackageDescription>
+<GeneratePackageOnBuild>true</GeneratePackageOnBuild>
+<ServicingVersion>1</ServicingVersion>
```
Comment (carlossanlop) [inline src/libraries/Microsoft.Extensions.Http/src/Microsoft.Extensions.Http.csproj:12]: nit:
```suggestion
    <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
    <ServicingVersion>1</ServicingVersion>
```

File: src/libraries/Microsoft.Extensions.FileProviders.Abstractions/src/Microsoft.Extensions.FileProviders.Abstractions.csproj [hunk 3]
```diff
@@ -10,6 +10,8 @@ Commonly Used Types:
 Microsoft.Extensions.FileProviders.IDirectoryContents
 Microsoft.Extensions.FileProviders.IFileInfo
 Microsoft.Extensions.FileProviders.IFileProvider</PackageDescription>
+<GeneratePackageOnBuild>true</GeneratePackageOnBuild>
+<ServicingVersion>1</ServicingVersion>
```
Comment (carlossanlop) [inline src/libraries/Microsoft.Extensions.FileProviders.Abstractions/src/Microsoft.Extensions.FileProviders.Abstractions.csproj:14]: nit:
```suggestion
    <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
    <ServicingVersion>1</ServicingVersion>
```

File: src/libraries/Microsoft.Extensions.Primitives/src/Microsoft.Extensions.Primitives.csproj [hunk 4]
```diff
@@ -11,6 +11,8 @@ Commonly Used Types:
 Microsoft.Extensions.Primitives.IChangeToken
 Microsoft.Extensions.Primitives.StringValues
 Microsoft.Extensions.Primitives.StringSegment</PackageDescription>
+    <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
+    <ServicingVersion>1</ServicingVersion>
```
Comment (carlossanlop) [inline src/libraries/Microsoft.Extensions.Primitives/src/Microsoft.Extensions.Primitives.csproj:15]: To anyone else reviewing this: The Microsoft.Extensions assemblies all have IsPackable set to true outside of the csproj.

**Review by carlossanlop — APPROVED** [review]
LGTM, just a couple of optional nits.
I'll write down the list of enabled OOB packages.

File: src/libraries/System.Memory.Data/src/System.Memory.Data.csproj (hunk 1 above)
Comment (ericstj) [reply src/libraries/System.Memory.Data/src/System.Memory.Data.csproj:6]: IsPackable doesn't mean it builds the package.  `GeneratePackageOnBuild` is what determines that in Runtime servicing.

Comment (carlossanlop) [reply src/libraries/System.Memory.Data/src/System.Memory.Data.csproj:6]: Gah yeah, sorry. Good point.

File: src/libraries/Microsoft.Extensions.Primitives/src/Microsoft.Extensions.Primitives.csproj (hunk 4 above)
Comment (ericstj) [reply src/libraries/Microsoft.Extensions.Primitives/src/Microsoft.Extensions.Primitives.csproj:15]: Yep - https://github.com/dotnet/runtime/blob/061c8b7177fbe56584994325a24bb262628ebd5a/src/libraries/Directory.Build.props#L79

**Comment by carlossanlop** [comment]
How's this looking @ericstj? Friendly reminder that today 10/14 is Code Complete for the November Release.

**Comment by ericstj** [comment]
Approved over email.

**Comment by ericstj** [comment]
This needs to resolve conflicts with https://github.com/dotnet/runtime/pull/108675

**Comment by ericstj** [comment]
@carlossanlop - Ok, this had some churn with resolving conflicts, but it should be ready to go.  The tests failing earlier were all known issues but BA gave up with too many wasm failures.

**Review by carlossanlop — APPROVED** [review]

**Comment by ericstj** [comment]
Both all configurations legs which build packages have completed.  Going to merge this to get the 6.0 servicing release moving

**Pull request merged**